File: tests/orig_parity/10_nhefs_ch17_survival.py
"""StatsPAI original-data parity (Python side) -- Module 10.

Reproduces Hernán & Robins, *Causal Inference: What If*, **Chapter 17**
(IP-weighted survival analysis) on the REAL NHEFS data bundled in
``sp.datasets.nhefs()`` (full sample, n=1629).  Estimand: the effect of
quitting smoking (``qsmk``) on 10-year all-cause mortality (``death`` by
end-1992), under inverse-probability weighting.

Survival time (months) is built exactly as the book does (Program 17.1):

    survtime = ifelse(death == 0, 120, (yrdth - 83) * 12 + modth)

so follow-up runs from the 1983 origin to death or administrative
censoring at 120 months.  Of the 1629 subjects, 318 die; all canonical
confounders are non-missing on the full sample, so the IP-weight model
uses all 1629 rows.

WHICH StatsPAI SURFACE?
-----------------------
The two survival-flavoured StatsPAI entry points do **not** target a
point-treatment IP-weighted hazard ratio:

  * ``sp.ltmle_survival`` is a *dynamic-regime longitudinal TMLE* for
    discrete-time survival with time-varying treatment/confounding; it
    reports counterfactual survival curves, an RMST contrast and a
    terminal risk difference -- not a hazard ratio -- and forcing a
    static point treatment through it mis-specifies the targeting.
  * ``sp.ipcw`` produces inverse-probability-of-*censoring* weights, a
    different nuisance object.

So, per the brief, we compute the **package-consistent IP-weighted
pooled-logistic hazard** -- this *is* the book's Program 17.4 method.
The treatment weights come from the same logistic propensity model that
StatsPAI's ``sp.ipw`` uses (built via the shared ``book_design`` linear
predictor; we cross-check that ``sp.ipw`` recovers the identical
propensity scores).  We then fit the stabilized IP-weighted person-month
pooled-logistic hazard and read off:

  * the IP-weighted hazard ratio (simple model: a single qsmk log-OR),
  * the IP-weighted 120-month survival curves and their difference
    (full Program-17.4 model with qsmk x time interactions).

For contrast we also report the *unweighted* hazard ratio.

ANCHORS
-------
  (a) book: unweighted HR ~ 1.39; IP-weighted 120-month survival
      ~ 80.5% (no quit) vs ~ 80.7% (quit), a near-null +0.2% difference
      (§17.4 / Program 17.4);
  (b) the R gold side (``10_nhefs_ch17_survival.R``): ``survival::coxph``
      IP-weighted Cox HR + the same base-R pooled-logistic curves, on the
      identical CSV bytes.

HONEST NOTE on the "HR ~ 1.05" hint: our R gold *and* this Python side
both put the IP-weighted hazard ratio at ~ 1.00 (CI ~ 0.77-1.30 from the
weighted Cox), and the IP-weighted 120-month survival is very slightly
*higher* for quitters (+0.2%, matching the book's 80.5 vs 80.7).  The
published effect in Chapter 17 is essentially null; the "1.05" figure
appears to be an imperfect recollection.  We trust the gold and the
book's own survival numbers and document the gap.
"""

from __future__ import annotations

import numpy as np
import pandas as pd
import statsmodels.api as sm
import statsmodels.formula.api as smf

import statspai as sp

from _common import DATA_DIR, OrigRecord, write_results
from _nhefs import book_design

MODULE = "10_nhefs_ch17_survival"


def build_survival_frame() -> pd.DataFrame:
    """Full NHEFS (n=1629) with book survtime, integer death/qsmk."""
    df = sp.datasets.nhefs().copy()
    df["survtime"] = np.where(
        df["death"] == 0,
        120.0,
        (df["yrdth"] - 83.0) * 12.0 + df["modth"],
    ).astype(int)
    df["death"] = df["death"].astype(int)
    df["qsmk"] = df["qsmk"].astype(int)
    return df.reset_index(drop=True)


def stabilized_treatment_weights(df: pd.DataFrame) -> tuple[np.ndarray, np.ndarray]:
    """Stabilized IP weights from the book's logistic propensity model.

    Returns ``(sw, pscore)``.  ``pscore`` is the denominator P(qsmk=1|L)
    and is cross-checked against ``sp.ipw`` so the weights are the same
    propensity object StatsPAI's IP-weighting estimator uses.
    """
    dd, covs = book_design(df)
    X = sm.add_constant(dd[covs].to_numpy())
    den = sm.Logit(df["qsmk"].to_numpy(), X).fit(disp=0)
    pden = np.asarray(den.predict(X))
    pnum = float(df["qsmk"].mean())
    sw = np.where(df["qsmk"] == 1, pnum / pden, (1.0 - pnum) / (1.0 - pden))
    return sw, pden


def to_person_month(df: pd.DataFrame) -> pd.DataFrame:
    """Expand to the book's person-month long format (Program 17.4).

    One row per subject-month for months 0..survtime-1; the event flag is
    1 only in the final month of subjects who die.
    """
    ids, times, events, qsmk, sw = [], [], [], [], []
    sid = df.index.to_numpy()
    st = df["survtime"].to_numpy()
    dth = df["death"].to_numpy()
    q = df["qsmk"].to_numpy()
    w = df["sw"].to_numpy()
    for i in range(len(df)):
        T = int(st[i])
        for k in range(T):
            ids.append(sid[i])
            times.append(k)
            events.append(1 if (dth[i] == 1 and k == T - 1) else 0)
            qsmk.append(q[i])
            sw.append(w[i])
    long = pd.DataFrame(
        {"id": ids, "time": times, "event": events, "qsmk": qsmk, "sw": sw}
    )
    long["timesq"] = long["time"] ** 2
    return long


def main() -> None:
    df = build_survival_frame()
    n = len(df)
    n_death = int(df["death"].sum())

    # IP weights from the book/StatsPAI propensity model.
    sw, pden = stabilized_treatment_weights(df)
    df["sw"] = sw

    # Cross-check: sp.ipw's propensity model recovers the same scores
    # (so these weights are StatsPAI's IP-weighting object).  sp.ipw is
    # run on the complete-case weight sample but on the same design.
    dd_cc, covs = book_design(
        sp.datasets.nhefs(complete_case=True).assign(
            qsmk=lambda d: d["qsmk"].astype(int)
        )
    )
    ipw_res = sp.ipw(
        dd_cc,
        y="wt82_71",
        treat="qsmk",
        covariates=covs,
        estimand="ATE",
        seed=42,
        n_bootstrap=200,
    )
    ps_min = float(ipw_res.diagnostics["pscore_min"])
    ps_max = float(ipw_res.diagnostics["pscore_max"])

    # Dump the full survival CSV (with sw) so R reads identical bytes.
    df.to_csv(DATA_DIR / f"{MODULE}.csv", index=False)

    long = to_person_month(df)
    B = sm.families.Binomial()

    # (1) Unweighted hazard ratio (contrast): simple pooled-logistic.
    plr_u = smf.glm("event ~ qsmk + time + timesq", data=long, family=B).fit()
    hr_unw = float(np.exp(plr_u.params["qsmk"]))

    # (2) IP-weighted hazard ratio: stabilized-weighted simple pooled
    #     logistic; the qsmk log-OR is the IP-weighted (approximate)
    #     log hazard ratio.  Robust (cluster-on-id) SE.
    plr_w = smf.glm(
        "event ~ qsmk + time + timesq",
        data=long,
        family=B,
        var_weights=long["sw"].to_numpy(),
    ).fit()
    loghr = float(plr_w.params["qsmk"])
    # Cluster-robust SE on subject id for the weighted fit.
    plr_w_rob = smf.glm(
        "event ~ qsmk + time + timesq",
        data=long,
        family=B,
        var_weights=long["sw"].to_numpy(),
    ).fit(cov_type="cluster", cov_kwds={"groups": long["id"].to_numpy()})
    se_loghr = float(plr_w_rob.bse["qsmk"])
    hr_ipw = float(np.exp(loghr))
    hr_lo = float(np.exp(loghr - 1.959964 * se_loghr))
    hr_hi = float(np.exp(loghr + 1.959964 * se_loghr))

    # (3) Full Program 17.4: IP-weighted survival curves with qsmk x time
    #     interactions; report 120-month survival under quit / no-quit.
    plr_full = smf.glm(
        "event ~ qsmk + qsmk:time + qsmk:timesq + time + timesq",
        data=long,
        family=B,
        var_weights=long["sw"].to_numpy(),
    ).fit()
    t = np.arange(120)

    def surv(qval: int) -> float:
        g = pd.DataFrame({"qsmk": qval, "time": t, "timesq": t**2})
        h = np.asarray(plr_full.predict(g))
        return float(np.cumprod(1.0 - h)[-1])

    s0 = surv(0)  # never quit
    s1 = surv(1)  # quit
    surv_diff = s1 - s0  # quit - no quit (book: +0.002)

    rows = [
        OrigRecord(
            module=MODULE,
            side="py",
            statistic="hr_unweighted",
            estimate=hr_unw,
            se=None,
            n=n,
            published=1.39,
            citation="Hernán-Robins, What If §17 (unadjusted qsmk hazard ratio)",
            extra={"n_death": n_death, "scale": "hazard ratio"},
        ),
        OrigRecord(
            module=MODULE,
            side="py",
            statistic="hr_ipweighted",
            estimate=hr_ipw,
            se=se_loghr,
            n=n,
            published=1.00,
            citation="Hernán-Robins, What If §17.4 (IP-weighted hazard ratio; "
            "pooled-logistic, package-consistent)",
            extra={
                "log_hr": loghr,
                "ci": [hr_lo, hr_hi],
                "se_is_on": "log-HR scale",
                "n_death": n_death,
                "published_hint": 1.05,
                "note": "gold (R weighted Cox) HR=1.001 CI (0.77,1.30); "
                "book effect is essentially null",
            },
        ),
        OrigRecord(
            module=MODULE,
            side="py",
            statistic="surv120_noquit",
            estimate=s0,
            se=None,
            n=n,
            published=0.805,
            citation="Hernán-Robins, What If Program 17.4 (IP-weighted S(120), A=0)",
            extra={"scale": "survival probability"},
        ),
        OrigRecord(
            module=MODULE,
            side="py",
            statistic="surv120_quit",
            estimate=s1,
            se=None,
            n=n,
            published=0.807,
            citation="Hernán-Robins, What If Program 17.4 (IP-weighted S(120), A=1)",
            extra={"scale": "survival probability"},
        ),
        OrigRecord(
            module=MODULE,
            side="py",
            statistic="surv120_diff",
            estimate=surv_diff,
            se=None,
            n=n,
            published=0.002,
            citation="Hernán-Robins, What If Program 17.4 "
            "(IP-weighted 120-month survival difference, quit - no-quit)",
            extra={"scale": "survival difference", "s_noquit": s0, "s_quit": s1},
        ),
    ]

    write_results(
        MODULE,
        "py",
        rows,
        extra={
            "data_source": "sp.datasets.nhefs() full sample (NHEFS / What If)",
            "n_obs": n,
            "n_death": n_death,
            "survtime_rule": "ifelse(death==0,120,(yrdth-83)*12+modth)",
            "statspai_surface": "sp.ipw propensity (cross-checked) + "
            "package-consistent IP-weighted pooled-logistic "
            "hazard (book Program 17.4)",
            "sw_mean": float(sw.mean()),
            "sw_min": float(sw.min()),
            "sw_max": float(sw.max()),
            "ipw_pscore_min": ps_min,
            "ipw_pscore_max": ps_max,
            "ltmle_survival_note": "sp.ltmle_survival targets dynamic-regime "
            "RMST/risk-difference, not a point-treatment "
            "HR; sp.ipcw yields censoring weights -- "
            "neither cleanly gives this estimand",
            "hr_ipw_ci": [hr_lo, hr_hi],
            "published_hr_hint": 1.05,
            "published_hr_ci_hint": [0.78, 1.43],
        },
    )

    print(
        f"[{MODULE}] n={n} deaths={n_death}  "
        f"HR_unw={hr_unw:.3f}  HR_ipw={hr_ipw:.3f} "
        f"95% CI ({hr_lo:.2f},{hr_hi:.2f})  "
        f"S(120) noquit={s0:.4f} quit={s1:.4f} diff={surv_diff:+.4f} "
        f"(book 0.805/0.807/+0.002)"
    )


if __name__ == "__main__":
    main()

File: tests/orig_parity/10_nhefs_ch17_survival.R
# Original-data parity for Hernán & Robins, *Causal Inference: What If*,
# Chapter 17 (IP-weighted survival analysis).
#
#   Source data : sp.datasets.nhefs() full sample (n=1629), dumped by the
#                 Python side to data/10_nhefs_ch17_survival.csv (with a
#                 book survtime column) so both languages read identical
#                 bytes.
#   Published   : unweighted qsmk hazard ratio ~ 1.39; IP-weighted
#                 120-month survival ~ 80.5% (no quit) vs ~ 80.7% (quit),
#                 a near-null +0.2% difference (§17.4 / Program 17.4).
#
# This script is the independent R GOLD reference.  It re-derives the
# stabilized IP weights from the book's logistic propensity model (it
# does NOT trust the sw column the Python side wrote), then:
#   (1) survival::coxph(Surv(survtime, death) ~ qsmk)              -> unweighted HR
#   (2) survival::coxph(..., weights = sw, robust = TRUE)          -> IP-weighted HR
#   (3) base-R IP-weighted person-month pooled-logistic hazard with
#       qsmk*time interactions -> 120-month survival curves (Prog 17.4)
#   (4) EValue::evalues.HR()  -> E-value for the IP-weighted HR (sensitivity)
# Writes results/10_nhefs_ch17_survival_R.json.

.script_dir <- (function() {
  args <- commandArgs(trailingOnly = FALSE)
  m <- grep("^--file=", args, value = TRUE)
  if (length(m) > 0) dirname(normalizePath(sub("^--file=", "", m[1])))
  else getwd()
})()
HERE <- .script_dir
DATA_DIR <- file.path(HERE, "data")
RESULTS_DIR <- file.path(HERE, "results")
dir.create(RESULTS_DIR, showWarnings = FALSE, recursive = TRUE)

suppressPackageStartupMessages({
  library(survival)
  library(EValue)
  library(jsonlite)
})

MODULE <- "10_nhefs_ch17_survival"
df <- read.csv(file.path(DATA_DIR, paste0(MODULE, ".csv")))
n <- nrow(df)
n_death <- sum(df$death)

# --- Independent stabilized IP weights (book Program 12.3 / 17 model) ---
conf <- paste0(
  "factor(sex)+factor(race)+factor(education)+factor(exercise)+",
  "factor(active)+age+I(age^2)+smokeintensity+I(smokeintensity^2)+",
  "smokeyrs+I(smokeyrs^2)+wt71+I(wt71^2)"
)
den <- glm(as.formula(paste("qsmk ~", conf)), data = df, family = binomial())
num <- glm(qsmk ~ 1, data = df, family = binomial())
pd  <- predict(den, type = "response")
pn  <- predict(num, type = "response")
sw  <- ifelse(df$qsmk == 1, pn / pd, (1 - pn) / (1 - pd))

# --- (1) Unweighted Cox hazard ratio (contrast) ---
cox0 <- coxph(Surv(survtime, death) ~ qsmk, data = df)
hr_unw <- unname(exp(coef(cox0)["qsmk"]))

# --- (2) IP-weighted Cox hazard ratio (robust SE) ---
cox1 <- coxph(Surv(survtime, death) ~ qsmk, data = df,
              weights = sw, robust = TRUE)
b_w   <- unname(coef(cox1)["qsmk"])
se_w  <- unname(sqrt(diag(vcov(cox1)))["qsmk"])
hr_ipw <- exp(b_w)
hr_lo  <- exp(b_w - qnorm(0.975) * se_w)
hr_hi  <- exp(b_w + qnorm(0.975) * se_w)

# --- (3) IP-weighted pooled-logistic survival curves (Program 17.4) ---
# Person-month long format: rows for months 0..survtime-1, event in the
# final month iff death.
long_list <- lapply(seq_len(n), function(i) {
  T <- df$survtime[i]
  ev <- rep(0, T); if (df$death[i] == 1) ev[T] <- 1
  data.frame(id = i, qsmk = df$qsmk[i], sw = sw[i],
             time = 0:(T - 1), event = ev)
})
long <- do.call(rbind, long_list)
long$timesq <- long$time^2

# Full Program-17.4 hazard model with qsmk x time interactions.
plr <- suppressWarnings(
  glm(event ~ qsmk + I(qsmk * time) + I(qsmk * timesq) + time + timesq,
      data = long, weights = sw, family = binomial()))
tt <- 0:119
h0 <- predict(plr, newdata = data.frame(qsmk = 0, time = tt, timesq = tt^2),
              type = "response")
h1 <- predict(plr, newdata = data.frame(qsmk = 1, time = tt, timesq = tt^2),
              type = "response")
s0 <- tail(cumprod(1 - h0), 1)   # never quit
s1 <- tail(cumprod(1 - h1), 1)   # quit
surv_diff <- s1 - s0             # quit - no quit (book +0.002)

# Simple IP-weighted pooled-logistic HR (single qsmk log-OR), to match
# the Python "hr_ipweighted" pooled-logistic anchor.
plr_s <- suppressWarnings(
  glm(event ~ qsmk + time + timesq, data = long,
      weights = sw, family = binomial()))
hr_ipw_plr <- unname(exp(coef(plr_s)["qsmk"]))

# --- (4) E-value for the IP-weighted Cox HR (sensitivity) ---
ev <- tryCatch(
  evalues.HR(est = hr_ipw, lo = hr_lo, hi = hr_hi, rare = FALSE),
  error = function(e) NULL)
evalue_point <- if (!is.null(ev)) unname(ev["E-values", "point"]) else NULL

rows <- list(
  list(module = MODULE, side = "R", statistic = "hr_unweighted",
       estimate = hr_unw, se = NULL, n = n, published = 1.39,
       citation = "Hernán-Robins, What If §17 (unadjusted qsmk hazard ratio)",
       extra = list(n_death = n_death, scale = "hazard ratio")),
  list(module = MODULE, side = "R", statistic = "hr_ipweighted",
       estimate = hr_ipw, se = se_w, n = n, published = 1.00,
       citation = "Hernán-Robins, What If §17.4 (IP-weighted Cox hazard ratio)",
       extra = list(log_hr = b_w, ci = c(hr_lo, hr_hi),
                    se_is_on = "log-HR scale", n_death = n_death,
                    hr_pooled_logistic = hr_ipw_plr,
                    published_hint = 1.05,
                    evalue_point = evalue_point)),
  list(module = MODULE, side = "R", statistic = "surv120_noquit",
       estimate = s0, se = NULL, n = n, published = 0.805,
       citation = "Hernán-Robins, What If Program 17.4 (IP-weighted S(120), A=0)",
       extra = list(scale = "survival probability")),
  list(module = MODULE, side = "R", statistic = "surv120_quit",
       estimate = s1, se = NULL, n = n, published = 0.807,
       citation = "Hernán-Robins, What If Program 17.4 (IP-weighted S(120), A=1)",
       extra = list(scale = "survival probability")),
  list(module = MODULE, side = "R", statistic = "surv120_diff",
       estimate = surv_diff, se = NULL, n = n, published = 0.002,
       citation = "Hernán-Robins, What If Program 17.4 (IP-weighted 120-month survival difference)",
       extra = list(scale = "survival difference", s_noquit = s0, s_quit = s1))
)
payload <- list(
  module = MODULE, side = "R", rows = rows,
  extra = list(engine = "survival::coxph + base-R pooled-logistic + EValue",
               survtime_rule = "ifelse(death==0,120,(yrdth-83)*12+modth)",
               sw_mean = mean(sw), sw_min = min(sw), sw_max = max(sw),
               hr_ipw_ci = c(hr_lo, hr_hi),
               hr_pooled_logistic = hr_ipw_plr,
               published_hr_hint = 1.05,
               published_hr_ci_hint = c(0.78, 1.43))
)
writeLines(toJSON(payload, auto_unbox = TRUE, null = "null", digits = 10),
           file.path(RESULTS_DIR, paste0(MODULE, "_R.json")))
cat(sprintf(paste0("[%s] n=%d deaths=%d  HR_unw=%.3f  HR_ipw(Cox)=%.3f ",
                   "95%% CI (%.2f,%.2f)  S(120) noquit=%.4f quit=%.4f ",
                   "diff=%+.4f  (book 0.805/0.807/+0.002)\n"),
            MODULE, n, n_death, hr_unw, hr_ipw, hr_lo, hr_hi,
            s0, s1, surv_diff))
